sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamingSymmetricHashJoinExec.scala

```diff
@@ -0,0 +1,330 @@
+ *     a condition ~ "left < rightTime + Y" with left side event watermark is used to calculate
+ *     right side state watermark.
+ *   - Depending on the conditions, the state watermark maybe different for the left and right
+ *     side. In the above example, leftTime > 12:26 AND rightTime > 12:34 - 1 hour = 11:34.
+ *   - State can be dropped from BOTH sides only when there are conditions of the above forms that
+ *     define time bounds on timestamp in both directions.
+ *
+ * 3. When both window in join key and time range condiions are present, case 1 + 2.
+ *    In this case, since window equality is a stricter condition than the time range, we can
+ *    use the the State Key Watermark = event time watermark to discard state (similar to case 1).
+ */
+case class StreamingSymmetricHashJoinExec(
+    leftKeys: Seq[Expression],
+    rightKeys: Seq[Expression],
+    joinType: JoinType,
+    condition: Option[Expression],
+    stateInfo: Option[StatefulOperatorStateInfo],
+    eventTimeWatermark: Option[Long],
+    stateWatermarkPredicates: JoinStateWatermarkPredicates,
+    left: SparkPlan,
+    right: SparkPlan) extends SparkPlan with BinaryExecNode with StateStoreWriter {
+
+  def this(
+      leftKeys: Seq[Expression],
+      rightKeys: Seq[Expression],
+      joinType: JoinType,
+      condition: Option[Expression],
+      left: SparkPlan,
+      right: SparkPlan) = {
+    this(
+      leftKeys, rightKeys, joinType, condition, stateInfo = None, eventTimeWatermark = None,
+      stateWatermarkPredicates = JoinStateWatermarkPredicates(), left, right)
+  }
+
+  assert(joinType == Inner, s"${getClass.getSimpleName} should not take $joinType as the JoinType")
+  assert(leftKeys.map(_.dataType) == rightKeys.map(_.dataType))
+
+  private val storeConf = new StateStoreConf(sqlContext.conf)
+  private val hadoopConfBcast = sparkContext.broadcast(
+    new SerializableConfiguration(SessionState.newHadoopConf(
+      sparkContext.hadoopConfiguration, sqlContext.conf)))
+
+  /**
+   * Join keys of both sides generate rows of the same fields, that is, same sequence of data types.
+   * If one side (say left side) has a column (say timestmap) that has a watermark on it,
+   * then it will never consider joining keys that are < state key watermark (i.e. event time
+   * watermark). On the other side (i.e. right side), even if there is no watermark defined,
+   * there has to be an equivalent column (i.e. timestamp). And any right side data that has the
+   * timestamp < watermark will not match will not match with left side data, as the left side get
+   * filtered with the explicitly defined watermark. So, the watermark in timestamp column in
+   * left side keys effectively causes the timestamp on the right side to have a watermark.
+   * We will use the ordinal of the left timestamp in the left keys to find the corresponding
+   * right timestamp in the right keys.
+   */
+  private val joinKeyOrdinalForWatermark: Option[Int] = {
```

Comment:
how does this behave if I union two streams with the same schema, but one stream has a watermark delay of 25 minutes whereas the other one has a delay of 5 minutes.

Reply:
That's a separate problem with unions. I am not sure what we do wither watermarks when we join. It could be that we dont preserve the watermark metadata. Out of scope for this PR.